04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+        "\n",
+        "print(reduce(lambda acc, number: acc * number, numbers))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "iDUnY5Ikg1nA",
+        "outputId": "7348d1d5-94e8-4896-d744-a1755b37af78"
+      },
+      "execution_count": 56,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "120\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "А можем и да подадем начална стойност:"
+      ],
+      "metadata": {
+        "id": "Y0kpWGxPrm1B"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "numbers = [3, 5, 8]\n",
+        "print(reduce(lambda acc, number: acc * number, numbers, 1))\n",
+        "print(reduce(lambda acc, number: acc * number, numbers, 2))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "l1W8u5RrrpyN",
+        "outputId": "9d1374c3-48a0-4674-cc95-37f8fd40f408"
+      },
+      "execution_count": 57,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "120\n",
+            "240\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "А сега един по-сложен пример: Нека имаме списък, чийто елементи са списъци с числа. С помощта на `reduce`, подходяща функция и подходяща начална стойност, бихме могли да получим един списък, който съдържа числата, без те да са в отделни списъци."
+      ],
+      "metadata": {
+        "id": "YVMFH9kZsOkV"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "numbers_2d = [[1, 2, 3], [4, 5, 6], [7, 8, 9, 10]]\n",
+        "\n",
+        "print(reduce(lambda acc, item: acc + item, numbers_2d, []))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "QQn2gsO-snB-",
+        "outputId": "de0281c4-a741-4005-a076-b6a6be898a39"
+      },
+      "execution_count": 58,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Припомням, че операция събиране (`+`) приложена върху два списъка ни връща конкатенацията на двата списъка. \n",
+        "\n",
+        "В горната операция, ние взехме един списък, и го \"изравнихме\" (flatten). Операцията \"flatten\" върху списъци е често срещата в езиците за функционално програмиране, макар и в Python да нямаме готова функция за това. "
+      ],
+      "metadata": {
+        "id": "0S1DKM_ps2jd"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Zip"
```

Comment:
There's also `sum`, `max`, `min`, `all`, `any`

Reply:
Don't think this is the place to include them.

Reply:
or at least to put them in a list without giving examples for each of them, just so students know that they exist

Reply:
Let's discuss on the next meeting